### PR TITLE
Integrate user registration between client-side and server-side

### PR DESCRIPTION
**Pull request by Oleh-Stefanovych**
This PR adds client/server integration of the registration process.

`WebApplication` represents a Spark-based web server that deploys service on `localhost:5001`.

`HttpCommunicationOptions` configures communication options available for the web application. All requests to the server are filtered with `RequestLoggerFilter`. `CORS` restrictions are configured with `CorsFilter`.

`RegisterUserRoute` handles `POST` requests to the `/api/register` endpoint for processing user registration in the FileHub system.

The `Spark` version has been updated to `2.9.1`.

### CODE REVIEW

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import com.google.gson.JsonSerializer;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * The facade of the {@link Gson} utility, which provides the conversion of an {@link Object} to an
+ * equivalent JSON representation.
+ *
+ * <p>To build a JSON representation of an object, a custom {@linkplain JsonSerializer serializer}
+ * is used that defines the serialization rules.
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java]: `A custom {@linkplain JsonSerializer serializer}, which defines the serialization rules, is used to build a JSON representation of an object`.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @param <T>
+ *         the class of the object to be converted
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java]: `a type of object to be converted`.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonSyntaxException;
+
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * The facade of the {@link Gson} utility, which provides the conversion of JSON to an equivalent
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: `JSON object`

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ * instance.
+ *
+ * <p>To convert an existing JSON string to an equivalent instance, a custom {@linkplain
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: which instance?

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * JsonDeserializer deserializer} is used that defines the deserialization rules.
+ *
+ * @param <T>
+ *         the class of the object to be created
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: `a type of object to be converted`.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+     *         the {@linkplain JsonDeserializer deserializer} that defines the deserialization rules
+     *         for the converted object
+     */
+    public JsonParser(Class<T> clazz, D deserializer) {
+        this.clazz = checkNotNull(clazz);
+        this.deserializer = checkNotNull(deserializer);
+    }
+
+    /**
+     * Deserializes the JSON string into an equivalent object.
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: so `object` or `instance`?

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     * <p>Note that this method uses {@link Gson#fromJson(String, Class)} for deserialization.
+     * The {@link Gson} utility is created and configured by the passed deserializer.
+     *
+     * @param json
+     *         the JSON string
+     * @return the converted from JSON string object instance
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: `the object instance converted from the JSON string`

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+     * @throws JsonSyntaxException
+     *         if JSON is not a valid representation for an object
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: `JSON object`

File: server/src/main/java/io/javaclasses/filehub/web/json/package-info.java [hunk 9]
```diff
@@ -0,0 +1,4 @@
+/**
+ * This package describes utilities for working with JSON.
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/json/package-info.java]: `JSON objects`

File: server/src/main/java/io/javaclasses/filehub/web/package-info.java [hunk 10]
```diff
@@ -0,0 +1,4 @@
+/**
+ * This package describes web application routines for the FileHub cloud file storage.
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/package-info.java]: `file` is redundant. Check through the application.

File: web-client/app/service/request-service/index.js [hunk 11]
```diff
@@ -96,8 +96,8 @@ export class RequestService {
   }
 
   /**
-   * Handles a {@link Promise} of `fetch` method. Retuens a {@link Promise} which resolves with
-   * server response message or  rejects with server error message or runtime message.
+   * Handles a {@link Promise} of `fetch` method. Returns a {@link Promise} which resolves with
+   * server response message or rejects with server error message or runtime message.
```
Comment (denysLavrienov) [inline web-client/app/service/request-service/index.js]: `runtime error message`

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+ * {
+ *   "login": "<login_value>",
+ *   "password": "<password_value>",
+ *   "avatar": "<avatar_image_content>"
+ * }
+ * }</pre>
+ *
+ * where {@code login} and {@code password} values are non-encrypted user credentials,
+ * and {@code avatar} is user avatar image content encoded as Base64.
+ * All these members are mandatory to deserialize JSON string into {@link RegisterUser} instance.
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: Replaced.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+ */
+class RegisterUserDeserializer implements JsonDeserializer<RegisterUser> {
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Instance of {@link RegisterUser} may not be possible to create if the validation
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: Fixed.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+     * requirements not satisfied. See {@link RegisterUser#checkLogin(String)} and
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: Fixed.

File: server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Filter;
+import spark.Request;
+import spark.Response;
+
+/**
+ * A {@linkplain Filter filter} that logs the HTTP request details.
+ *
+ * <p>A log message will consist of request method and target URL.
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java]: Fixed.

File: server/src/main/java/io/javaclasses/filehub/web/WebApplication.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+import static spark.Spark.after;
+import static spark.Spark.options;
+import static spark.Spark.path;
+import static spark.Spark.post;
+
+/**
+ * A web application for the FileHub cloud file storage.
+ *
+ * <p>The server based on Spark Framework.
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/WebApplication.java]: Fixed.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java (hunk 1 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java]: Replaced.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java (hunk 2 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java]: Replaced.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java (hunk 3 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: Replaced with `JSON representation`.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java (hunk 6 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: `object instance` 😃

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java (hunk 7 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: Replaced.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java (hunk 8 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: Replaced with `if JSON representation is not equal to object structure`.

File: server/src/main/java/io/javaclasses/filehub/web/json/package-info.java (hunk 9 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/package-info.java]: Replaced with `JSON format.`.

File: server/src/main/java/io/javaclasses/filehub/web/package-info.java (hunk 10 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/package-info.java]: Fixed all occurrences.

File: web-client/app/service/request-service/index.js (hunk 11 above)
Comment (Oleh-Stefanovych) [reply web-client/app/service/request-service/index.js]: `or runtime message.` has been removed.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java (hunk 4 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonParser.java]: Fixed.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
+import static javax.servlet.http.HttpServletResponse.SC_CONFLICT;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+
+/**
+ * A {@linkplain Route route} which is designed to handle a user registration request.
+ *
+ * <p>The {@link Registration} process is used to register a user. The {@link RegisterUser} command
+ * is converted from the request body using {@link RegisterUserDeserializer}.
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: Removed.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+     * {@link RegisterUser#checkPassword(String)} for the implied requirements on the user
+     * credentials.
+     *
+     * @throws InvalidCredentialsException
+     *         if the user credentials are invalid
+     */
+    @Override
+    public
+    RegisterUser deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: The class has been marked as `final`.

File: web-client/app/service/registration-request-service/index.js [hunk 19]
```diff
@@ -15,15 +15,40 @@ export class RegistrationRequestService {
...
-    const formData = new FormData();
-    Object.keys(user).forEach((key) => {
-      formData.append(key, user[key]);
+    return this.encodeImageFileToBase64(user.avatar).then((avatarAsBase64) => {
+      const requestBody = JSON.stringify({
+        login: user.login,
+        password: user.password,
+        avatar: avatarAsBase64,
+      });
+
```
Comment (kuzmavictor) [inline web-client/app/service/registration-request-service/index.js]: Redundant line break.

File: web-client/app/service/registration-request-service/index.js [hunk 20]
```diff
@@ -15,15 +15,40 @@ export class RegistrationRequestService {
   /**
    * Sends a user registration request.
    *
-   * @param {User} user - an user registration credentials.
-   * @return {Promise} - a {@link Promise} which resolves if the request was successful or
-   * rejects with server error message or runtime message.
+   * @param {User} user - the user registration credentials
+   * @return {Promise} - the {@link Promise} which resolves if the request was successful or
+   * rejects with server error message
```
Comment (kuzmavictor) [inline web-client/app/service/registration-request-service/index.js]: Rewrite sentence it contains grammar errors. Check rules about conditions.

File: server/src/main/java/io/javaclasses/filehub/web/package-info.java [hunk 21]
```diff
@@ -0,0 +1,4 @@
+/**
+ * This package describes web application routines for the FileHub cloud storage.
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/package-info.java]: The current description is not correct. You try to implement a controller (REST) layer.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+     * <p>An instance of {@link RegisterUser} may not be possible to be created if the validation
+     * requirements are not satisfied. See ctor documentation of the {@link RegisterUser} for the
+     * implied requirements on the user credentials.
+     *
+     * @throws InvalidCredentialsException
+     *         if the user credentials are invalid
+     */
+    @Override
+    public
+    RegisterUser deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: `Type typeOfT`,` JsonDeserializationContext context` is redundant parameters that are never used.

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
+package io.javaclasses.filehub.web.json;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSerializer;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * The facade of the {@link Gson} utility, which provides the conversion of an {@link Object} to an
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java]: This is not a facade. Change the description.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+     * <p>Depending on the registration result, the response may have the following HTTP status
+     * codes:
+     *
+     * <ul>
+     *      <li>200 — if the registration was successful;
+     *      <li>400 — if the request body or user credentials are invalid;
+     *      <li>409 — if such a user already exists.
+     * </ul>
+     *
+     * @return the result of user registration in a JSON format defined in {@link MessageSerializer}
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: `the content to be set in the response`

File: server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * <p>A log message consists of request method and target URL.
+ */
+final class RequestLoggerFilter implements Filter {
+
+    private static final Logger logger = LoggerFactory.getLogger(RequestLoggerFilter.class);
+
+    @Override
+    public void handle(Request request, Response response) throws Exception {
+
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java]: Redundant line break.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+ *   "login": "<login_value>",
+ *   "password": "<password_value>",
+ *   "avatar": "<avatar_image_content>"
+ * }
+ * }</pre>
+ *
+ * where {@code login} and {@code password} values are non-encrypted user credentials,
+ * and {@code avatar} is user avatar image content encoded as Base64.
+ * All of these members are required to convert a JSON to a {@link RegisterUser} instance
+ */
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: Do not specify implementation. Describe why this is necessary.

File: server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+ * single string message to JSON.
+ *
+ * A serialized string into JSON will look like this:
+ *
+ * <pre>{@code
+ * {
+ *   "message": "<message_text>"
+ * }
+ * }</pre>
+ */
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java]: Do not specify implementation. Describe why this is necessary.

File: web-client/app/service/registration-request-service/index.js (hunk 19 above)
Comment (Oleh-Stefanovych) [reply web-client/app/service/registration-request-service/index.js]: Let's discuss this in person.

File: web-client/app/service/registration-request-service/index.js (hunk 20 above)
Comment (Oleh-Stefanovych) [reply web-client/app/service/registration-request-service/index.js]: Fixed.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java (hunk 22 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: This is an overridden method.

File: server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java (hunk 25 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java]: The line is not redundant. It improves the code readability.

File: server/src/main/java/io/javaclasses/filehub/web/package-info.java (hunk 21 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/package-info.java]: Updated.

File: server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java (hunk 27 above)
Comment (okucheruk) [reply server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java]: It's better to mark it as `@apiNote` to describe the contract obligates.
For more information see [here](https://bugs.openjdk.java.net/browse/JDK-8068562).

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java (hunk 26 above)
Comment (okucheruk) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: It's better to mark it as `@apiNote` to describe the contract obligates.
For more information see [here](https://bugs.openjdk.java.net/browse/JDK-8068562).

Comment (okucheruk) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: It's better to mark it as `@apiNote `to describe the contract obligates.
For more information see [here](https://bugs.openjdk.java.net/browse/JDK-8068562).

File: server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java (hunk 27 above)
Comment (okucheruk) [reply server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java]: It's better to mark it as `@apiNote `to describe the contract obligates.
For more information see [here](https://bugs.openjdk.java.net/browse/JDK-8068562).

File: server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java (hunk 23 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/json/JsonBuilder.java]: Updated,

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java (hunk 24 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: Marked as `apiNote`.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java (hunk 26 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: Discussed in person. Marked as `apiNote`.

File: server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java (hunk 27 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/MessageSerializer.java]: Discussed in person. Marked as `apiNote`.

File: web-client/app/service/registration-request-service/index.js (hunk 19 above)
Comment (Oleh-Stefanovych) [reply web-client/app/service/registration-request-service/index.js]: The line is not redundant. It improves the code readability.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+ *         {
+ *           "login": "<login_value>",
+ *           "password": "<password_value>",
+ *           "avatar": "<avatar_image_content>"
+ *         }
+ *         }</pre>
+ *
+ *         where {@code login} and {@code password} values are non-encrypted user credentials,
+ *         and {@code avatar} is user avatar image content encoded as Base64. All of these members
+ *         are required to convert a JSON to a {@link RegisterUser} instance.
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: Replaced with `All of these members are required.`.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+ * required.
+ */
+final class RegisterUserDeserializer implements JsonDeserializer<RegisterUser> {
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>An instance of {@link RegisterUser} may not be possible to be created if the validation
+     * requirements are not satisfied. See ctor documentation of the {@link RegisterUser} for the
+     * implied requirements on the user credentials.
```
Comment (kuzmavictor) [inline server/src/main/java/io/javaclasses/filehub/web/RegisterUserDeserializer.java]: You may tell it inside the exception description.  Also, the last sentence is redundant (you may use `@see`).  Please, describe the `JsonParseException`.

File: server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java (hunk 25 above)
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RequestLoggerFilter.java]: Куьщмувю

File: server/src/main/java/io/javaclasses/filehub/web/HttpCommunicationOptions.java [hunk 30]
```diff
@@ -0,0 +1,26 @@
+package io.javaclasses.filehub.web;
+
+import spark.Request;
+import spark.Response;
+import spark.Route;
+
+/**
+ * A {@linkplain Route route} which is intended to be mapped for OPTIONS HTTP request method.
+ * It allows the client to determine the options and/or requirements associated with a server.
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/HttpCommunicationOptions.java]: See [this](https://tools.ietf.org/html/rfc2616#section-9.2).

File: server/src/main/java/io/javaclasses/filehub/web/HttpCommunicationOptions.java [hunk 31]
```diff
@@ -0,0 +1,26 @@
...
+    /**
+     * {@inheritDoc}
+     *
+     * <p>The current implementation sets the value “*” to the {@code Access-Control-Allow-Headers}
+     * and {@code Access-Control-Allow-Methods} response headers.
+     */
+    @Override
+    public Object handle(Request request, Response response) {
+        response.header("Access-Control-Allow-Headers", "*");
+        response.header("Access-Control-Allow-Methods", "*");
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/HttpCommunicationOptions.java]: At this stage, I do not see the need to limit the allowed methods.

File: server/src/main/java/io/javaclasses/filehub/web/HttpCommunicationOptions.java [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>The current implementation sets the value “*” to the {@code Access-Control-Allow-Headers}
+     * and {@code Access-Control-Allow-Methods} response headers.
+     */
+    @Override
+    public Object handle(Request request, Response response) {
+        response.header("Access-Control-Allow-Headers", "*");
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/HttpCommunicationOptions.java]: At this stage, I do not see the need to limit the allowed headers.

File: server/src/main/java/io/javaclasses/filehub/web/WebApplication.java [hunk 33]
```diff
@@ -0,0 +1,54 @@
+package io.javaclasses.filehub.web;
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/WebApplication.java]: A class located in a nested package can also be part of the public API.

(c) Alex.

[the remark this answers is not in the document]
Comment (denysLavrienov) [reply server/src/main/java/io/javaclasses/filehub/web/WebApplication.java]: The root package contains classes that determine `How can I use it?`

(c) Oleg

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java [hunk 34]
```diff
@@ -0,0 +1,73 @@
...
+                new JsonBuilder<>(String.class, new MessageSerializer());
+        JsonParser<RegisterUser, RegisterUserDeserializer> registerUserParser =
+                new JsonParser<>(RegisterUser.class, new RegisterUserDeserializer());
+
+        String body = request.body();
+        response.type("application/json");
+
+        try {
+            RegisterUser registerUser = registerUserParser.parse(body);
+            Registration registration = new Registration(ServerEnvironment.instance()
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: `ServerEnvironment` was extracted as a constructor parameter.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java [hunk 35]
```diff
@@ -0,0 +1,73 @@
...
+                                                                          .userStorage());
+            registration.handle(registerUser);
+            response.status(SC_OK);
+            return responseBuilder.build("The user has been successfully registered.");
+
+        } catch (JsonSyntaxException e) {
+            response.status(SC_BAD_REQUEST);
+            return responseBuilder.build("Request body is invalid.");
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: Are you sure this is a good server response?
```JSON
{
    "message": "com.google.gson.stream.MalformedJsonException: Unterminated object at line 4 column 14 path $.avatar"
}
```

[the remark this answers is not in the document]
Comment (kuzmavictor) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: I sure exactly what you should be thinking and try to resolve problems before typing about failures or get more information if my comment is unclear. I want to see something like that. 
image

[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: “Let's not show SQL queries to housewives”
(c) @Oleg-Melnik

File: server/src/main/java/io/javaclasses/filehub/web/WebApplication.java (hunk 33 above)
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/WebApplication.java]: Let's discuss in person.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java (hunk 35 above)
[the remark this answers is not in the document]
Comment (kuzmavictor) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: This is not a good option for users. Also, you do not work with SQL, but submit the form to the server.
image

File: server/src/main/java/io/javaclasses/filehub/web/WebApplication.java (hunk 33 above)
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/WebApplication.java]: Discussed with @Oleg-Melnik. The current class structure is correct.

File: server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java (hunk 35 above)
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply server/src/main/java/io/javaclasses/filehub/web/RegisterUserRoute.java]: A `MemberNotFoundException` has been added.

If the JSON in the request body has an invalid format, the following response will be returned:
```JSON
{
    "message": "JSON request body has invalid syntax."
}
```

If the required field was not found, the following response will be returned:
```JSON
{
    "message": "The member `avatar` not found in the body of the JSON request."
}
```

File: server/src/main/java/io/javaclasses/filehub/web/WebApplication.java [hunk 36]
```diff
@@ -0,0 +1,55 @@
...
+import io.javaclasses.filehub.ServerEnvironment;
+import spark.Spark;
+
+import static spark.Spark.after;
+import static spark.Spark.options;
+import static spark.Spark.path;
+import static spark.Spark.post;
+
+/**
+ * An entry point for the FileHub cloud storage web application.
```
Comment (denysLavrienov) [inline server/src/main/java/io/javaclasses/filehub/web/WebApplication.java]: `FileHub application` is enough. `An entry point into ...`.

**Pull request merged**